### PR TITLE
Use boost::ranges for ObjectMap::all function

### PR DESCRIPTION
**Pull request by adrianbroher**
This PR:

* prevents the generation of an intermediate std::vector for ObjectMap::all calls and returns an boost range instead.
* Accepts a boost::any_range<int> as an input for the ObjectMap::find function to prevent code duplication.
* Drops the ObjectMap::{begin,end} interface and replaces it with the corresponding ObjectMap::{all,find} calls.

### CODE REVIEW

**Comment by adrianbroher** [comment]
ping

**Comment by geoffthemedio** [comment]
Why do you prefer to require `all()` instead of having that be the result of iterating over just `Objects()` ?

**Comment by adrianbroher** [comment]
It's inconsistent and cannot be templated. It's an eyesore.

File: python/UniverseWrapper.cpp [hunk 1]
```diff
@@ -76,9 +76,10 @@ namespace {
     std::vector<int> ObjectIDs(const Universe& universe)
     {
         std::vector<int> result;
-        result.reserve(universe.Objects().size());
-        for (auto it = universe.Objects().begin<T>(); it != universe.Objects().end<T>(); ++it)
-            result.push_back(it->ID());
+        auto objs = universe.Objects().all<T>();
```
Comment (geoffthemedio) [inline python/UniverseWrapper.cpp]: Is there a purpose or need for storing the result of calling all in a local rather than putting the all call into the loop declaration directly?

File: universe/UniverseGenerator.cpp [hunk 2]
```diff
@@ -658,9 +658,11 @@ void GenerateStarlanes(int max_jumps_between_systems, int max_starlane_length) {
     std::map<int, std::set<int>> potential_system_lanes;
 
     // get systems
-    auto sys_vec = Objects().all<System>();
+    auto sys_rng = Objects().all<System>();
+    std::vector<std::shared_ptr<System>> sys_vec;
```
Comment (geoffthemedio) [inline universe/UniverseGenerator.cpp]: sys_vec seems redundant now

File: util/ModeratorAction.cpp [hunk 3]
```diff
@@ -157,8 +157,7 @@ namespace {
     std::string GenerateSystemName() {
         static std::vector<std::string> star_names = UserStringList("STAR_NAMES");
 
-        const ObjectMap& objects = Objects();
-        std::vector<std::shared_ptr<const System>> systems = objects.all<System>();
+        auto systems = Objects().all<System>();
```
Comment (geoffthemedio) [inline util/ModeratorAction.cpp]: Can probably iterate over systems directly with `all<System>` below without this

File: python/UniverseWrapper.cpp (hunk 1 above)
Comment (adrianbroher) [reply python/UniverseWrapper.cpp]: I usually do those mass substitutions via `sed` and let the compiler decide if the change was good or not.  When I see some place for optimization when I check in I usually do so, but I miss sometimes stuff.

Comment (geoffthemedio) [reply python/UniverseWrapper.cpp]: I assumed something like that, but thought I should ask in case there was a reason for it still.

File: util/ModeratorAction.cpp (hunk 3 above)
Comment (adrianbroher) [reply util/ModeratorAction.cpp]: See d9fb4f9bf58fb845bf4547dd4b9e1f53bc7df8e0

File: python/UniverseWrapper.cpp (hunk 1 above)
Comment (adrianbroher) [reply python/UniverseWrapper.cpp]: See d9fb4f9bf58fb845bf4547dd4b9e1f53bc7df8e0

File: UI/MapWnd.cpp [hunk 4]
```diff
@@ -2753,13 +2753,37 @@ void MapWnd::InitTurn() {
     // connect system fleet add and remove signals
     for (auto& system : objects.all<System>()) {
         m_system_fleet_insert_remove_signals[system->ID()].push_back(system->FleetsInsertedSignal.connect(
-            boost::bind(&MapWnd::FleetsInsertedSignalHandler, this, _1)));
+            [this](const std::vector<std::shared_ptr<Fleet>>& fleets) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:2756]: is this relevant to this commit?

Comment (geoffthemedio) [reply UI/MapWnd.cpp:2756]: I don't object, but it seems out of place given the commit message...

Comment (adrianbroher) [reply UI/MapWnd.cpp:2756]: Yes, changing to boost::range would have required to update the method signature of the now-removed methods. As those were only called here I moved the logic into this place to avoid figuring out the correct signature for those methods.
